job-shop-scheduling/test_job_shop_scheduler.py

```diff
@@ -385,6 +386,30 @@ def test_get_jss_bqm(self):
         bqm = get_jss_bqm(jobs, max_time)
         self.assertIsInstance(bqm, BinaryQuadraticModel)
 
+    def test_stitch_kwargs(self):
+        """ Ensure stitch_kwargs is being passed through get_jss_bqm to dwavebinarycsp.stitch
+        """
+        jobs = {"sandwich": [("bread", 1), ("roast_beef", 1)],
+                "french_toast": [("egg", 1), ("bread", 1)]}
+        max_time = 3
+        impossible_stitch_kwargs = {"max_graph_size": 0}
+
+        # ImpossibleBQM should be raised, as the max_graph size is too small
+        self.assertRaises(ImpossibleBQM, get_jss_bqm, jobs, max_time, impossible_stitch_kwargs)
+
```

Comment:
The reason I didn't check a good stitch_kwargs is that I don't know how/if dwavebinarycsp.stitch defaults will change in the future, so there wasn't a way (that I could think of) to know if our passed args were being used or if the defaults were being used. I suppose a test case for JSP that is too large for the current defaults could be used, but for the above reasons I think that testing a bad/impossible would be more reliable.

I'm only now realizing from your proposal that I've completely neglected to consider min_classical_gap. Would I just be checking that a BQM is returned or would I also be checking some parameters of the BQM?

Reply:
Yes, I understand your concern about the possible future `dwavebinarycsp.stitch(..)` parameters. Your concern is valid, which is why all our demos have a `requirements.txt` file. It specifies the Ocean SDK version our particular demo will work with. Hence, the unit tests just need to work with the SDK version that we promised in `requirements.txt` (in our case, `dwave-ocean-sdk==1.4.0`).

tl;dr: the person bumping up the SDK version for this demo can update this unit test and deal with the new `stitch` in the future `dwave-ocean-sdk`.

In my suggested code for "good stitch kwargs", I was implicitly testing that no errors were raised. However, you could be more explicit and check that the `jss_bqm` actually exists. Perhaps the following might work:  `self.assertIsInstance(jss_bqm, dimod.BinaryQuadraticModel)`

Reply:
Two questions on this:
- Do you prefer to have both TestGetJssBqm and TestGetBqm? Since TestGetJssBqm calls get_bqm it essentially duplicates TestGetBqm.
- For the specific good_stitch_kwargs you proposed above, did you happen to try that? For me it fails at
```
 in insert_ising_model
    encoded_data['max_quadratic_bias'] = max(itervalues(quadratic))
ValueError: max() arg is an empty sequence
```
I haven't really used min_classical_gap before so I can't say if this is unusual or not, but the sample code in the stitch documentation where this parameter is set ran with no problems.

Was there any reason you included min_classical_gap, or was it just by chance?

Reply:
On `TestGetJssBqm` and `TestGetBqm`:
- Yes, I would agree with the duplication. I would prefer to keep `TestGetJssBqm.test_stitch_kwargs()` as it directly tests the function that most users will be using, `get_jss_bqm()`

On `good_stitch_kwargs`:
- `min_classical_gap` is the energy gap between the ground state and the next excited state. My request that each constraint in `stitch` has at least an energy gap of 3 was too aggressive (i.e. no equation was found that would satisfy this request). Lowering this gap to 1.5 will work. (I'm not using 2 because it's the default gap for `stitch(..)`).
- I had suggested `min_classical_gap` as it would be a reasonable kwarg that someone would want to change in stitch.

Reply:
FYI, I have made an issue to return a nicer error for the initial "bad" `good_stitch_kwargs`. dwavesystems/penaltymodel/issues/97

Reply:
My latest commit should wrap up the tests, if not, let me know what's left.

Thanks for making the issue, hopefully it will help more than just me.